### ISSUE TITLE
test mIoU

### ISSUE THREAD

**Issue by Linbooo** [JindongJiang/RedNet#6]
Hello,
     I have a pretty slow GPU. It will probably takes weeks for me to train all 1500 epochs :( So far I have trained 300 epochs, got test mIoU 45.6 which is far from yours (47.8), using my own test script. I want to ask if my result is reasonable, can it be near 47.8 when all epochs are done? or I have to do some other tricks to reach there such as multiscale evaluation? By the way, I did not take the unlabeled class into consideration when I calculated the test mIoU. I would appreciate it if you can share with us your test script and metric calculations :)) Thanks for the project code anyway :+1: 

/Linbo

**Comment by JindongJiang** [comment]
Hi Linbo,

If you are using the same batchsize as in the paper, the highest mIoU should be around 400 - 480. No other trick is required to achieve this result, but I did search the best result in the saved checkpoints.

Also, I will release the metric calculation code later.

Best,

Jindong

**Comment by Linbooo** [comment]
Thanks! have you tried using larger base learning rate? I do not have a powerful GPU cluster and the process is too slow for me :(( 感谢了

**Comment by Linbooo** [comment]
or have you tried to use 'poly' learning rate schedule instead of 'step'?

**Comment by JindongJiang** [comment]
I haven't tried these two, but you can certainly try using them.

**Comment by Linbooo** [comment]
Hi again. I have now trained 500 epochs without any stop. I tested each 5th epoch between 400 and 500 on test set, i.e. 405, 410, 415... The mIoU are always between 45.0 to 45.6. Im pretty sure that my metric is correct since it got 47.8 on the weights provided by you. The training used the exact default setting in your code. Should I let it train some more days, what do you think

**Comment by JindongJiang** [comment]
Hi, what about now? Still 45.6?

**Comment by Linbooo** [comment]
Sadly, no process at all. Now I have trained to 1100 epochs. All results after 500 epochs got worse than before. I think the highest mIoU I have now is 45.6 so far.  I think I need to do it like you wrote in the paper, which is checking the validation loss as you have trained for a certain period. Then picking up that checkpoint which has lowest val loss and reduce the learning rate, and continue to train. Loop this process many times. Is this approach you have used??

**Comment by JindongJiang** [comment]
That makes sense, results after 500 are basically overfitted. And Yes, I suggest you do that also. By the way, are you using PyTorch 0.4 or 1.0? I remember that the weight initialization of PyTorch had changed from 0.4 to 1.0?

**Comment by Linbooo** [comment]
I used 0.4.1, I think this is what you've used right?  I just found that you used batch-size 5 for the training, but you wrote "default=10" in help. What is the actual batch-size you used?

**Comment by JindongJiang** [comment]
Yes, that is correct. And I use 5 as the training batch size, sorry for the misleading help message.

**Comment by hxwork** [comment]
Hello，
I trained the model use pytorch 0.4.0 and the original settings. After 220 epochs I test the model use my own evaluate codes on the validation datasets , but the mIOU result can only achieve 32.0. I don't understand why the point is so low....I sincerely hope that you can release the evaluate code soon. Thanks for the project code anyway!!!

**Comment by Linbooo** [comment]
32.0 seems too low. I can provide you my evaluation code for both mIoU and Accuracy. I have tested it on the weights given by JindongJiang, and I got correct values for both metrics. Not sure if it is allowed to post code here. Can you give me your email

**Comment by hxwork** [comment]
That's owsome! my email is 851960956@qq.com. Thank you very much!!!

**Comment by hxwork** [comment]
I use your test code and the mIOU result is 42.63! My code may have some mistakes. Thank you!

**Comment by Linbooo** [comment]
nice nice! I think I got similar results as yours. As I said, the best one is 45.6 I got. Also as JindongJiang suggested, we need to do the exhaustive search in the validation loss to obtain 47.8, which I still doubt the validity, to be honest. It will take forever for my computer to train and test. If you ever reach 47.8, tell me. :))

**Comment by hxwork** [comment]
OK, I will try my best to achieve the best result as the author wrote in the paper. And if I make a breakthrough, I will tell you immediately!

**Comment by Linbooo** [comment]
thanks :+1:

**Comment by JindongJiang** [comment]
Thank you, guys. I'll be back helping you once I finish my finals.

**Comment by hxwork** [comment]
Hi, Llinbo,
I used batch size 8 and evaluated the model after training 300 epochs. And the result shows that mIoU is similar to the result of 135 epochs model, sometimes even worse than that. The highest mIoU I have achieved is 44.97, and I found that the download link of the pre-trained weights is invalid now, I will be very appreciate if you can share the pre-trained weights with me...

**Comment by Linbooo** [comment]
yooo. it's more than 300MB, you have google drive or something where I can share it?

**Comment by JindongJiang** [comment]
@stephenXh Hi, actually, I checked the pre-train file link, it is downloadable.

**Comment by zhouqi946** [comment]
> Hello,
> I have a pretty slow GPU. It will probably takes weeks for me to train all 1500 epochs :( So far I have trained 300 epochs, got test mIoU 45.6 which is far from yours (47.8), using my own test script. I want to ask if my result is reasonable, can it be near 47.8 when all epochs are done? or I have to do some other tricks to reach there such as multiscale evaluation? By the way, I did not take the unlabeled class into consideration when I calculated the test mIoU. I would appreciate it if you can share with us your test script and metric calculations :)) Thanks for the project code anyway 👍
> 
> /Linbo

训练时，交叉熵计算出现-1越界，你知道是什么原因吗？数据集的制作时按照作者的Readme完成的

**Comment by Rosalindjiewen** [comment]
Hello, thank you for sharing, which is very helpful to me. I would like to ask: where can I download metric calculation code? Thank you and look forward to your reply.

**Comment by JindongJiang** [comment]
Hi Rosalindjiewen, 

Currently, I don't have a plan to release the metric calculation code. However, you can check [this code snippet](https://github.com/zhixuan-lin/G-SWM/blob/1423553c30e4340cbe2278639316941cfd16acb6/src/model/gswm/fg.py#L1050) from another project we worked on. This should help in implementing it in this project.

**Comment by Rosalindjiewen** [comment]
@JindongJiang Hello, I am very happy to receive your reply and the link you shared. If there are any problems in the learning process, I will ask you again, thank you very much. ~

**Comment by oceanners** [comment]
Hello, the author. I want to ask why the label.npy file is missing from the dataset I downloaded from the official website

**Comment by JindongJiang** [comment]
Hi @oceanners, I actually don't remember where the `label.npy` is from, but it is highly possible that it is generated in a preprocessing step to make the label easier to access. Since the preprocessing code has been lost, I believe a better way is to implement an eval script using the original label and [this code snippet](https://github.com/zhixuan-lin/G-SWM/blob/1423553c30e4340cbe2278639316941cfd16acb6/src/model/gswm/fg.py#L1050).

**Comment by oceanners** [comment]
> 你好@oceanners，我实际上不记得`label.npy`它来自哪里，但它很可能是在预处理步骤中生成的，以使标签更易于访问。由于预处理代码已经丢失，我认为更好的方法是使用原始标签和[这段代码片段](https://github.com/zhixuan-lin/G-SWM/blob/1423553c30e4340cbe2278639316941cfd16acb6/src/model/gswm/fg.py#L1050)实现一个 eval 脚本。

First of all, thank you for your reply. I have a few questions about the split network of integrated depth information. Could you leave a contact information, such as QQ or WeChat? Thank you

**Comment by oceanners** [comment]
> 32.0 似乎太低了。我可以为您提供我的 mIoU 和准确性评估代码。我已经在 J​​indongJiang 给出的权重上对其进行了测试，并且我得到了两个指标的正确值。不确定是否允许在此处发布代码。你能给我你的电子邮件吗

hello can i have a code？ 1120199090@qq.com thank you

**Comment by amyzxq** [comment]
> 那是欠的！我的电子邮件 [851960956@qq.com](mailto:851960956@qq.com)。谢谢！！！

hello Can I get the test code？my email：38853612@qq.com

**Comment by zhangcj13** [comment]
@JindongJiang  hello author, could you share the eval code? also  the pre-trained weight URL link is not available，could you release it again？

**Comment by zhangcj13** [comment]
@Linbooo  Hello， can you provide your eval code？My email：[zhangcj13@126.com](url)， thank you very much！！！

**Comment by Magic-Pony** [comment]
> 32.0 似乎太低了。我可以为您提供我的 mIoU 和 Accuracy 评估代码。我已经根据 JindongJiang 给出的权重对其进行了测试，并且我得到了两个指标的正确值。不确定是否允许在此处发布代码。您能给我您的电子邮件吗

Hello, can you share me the evaluate code? That would help me a lot! Thank you~
My email: 1141509788@qq.com
